### PR TITLE
Enable Zsh completions

### PR DESCRIPTION
**Pull request by j-g00da**

### CODE REVIEW

File: src/duty/cli.py [hunk 1]
```diff
@@ -275,7 +278,18 @@ def main(args: list[str] | None = None) -> int:
...
+            shell = os.path.basename(os.environ.get('SHELL'))
+        else:
+            shell = opts.completion.lower()
+
+        if shell == 'zsh':
+            print(Path(__file__).parent.joinpath("completions.zsh").read_text())
+        elif shell == 'bash':
+            print(Path(__file__).parent.joinpath("completions.bash").read_text())
+        else:
+            raise NotImplementedError(f"Completion is only supported on Bash and Zsh, got '{shell}'.")
```
Comment (bswck) [inline src/duty/cli.py:291]:
```suggestion
        try:
            print(Path(__file__).parent.joinpath(f"completions.{shell}").read_text())
        except FileNotFoundError as exc:
            msg = f"Completion is only supported on Bash and Zsh, got '{shell}'."
            raise NotImplementedError(msg) from exc
```

so we don't need to touch this code upon adding support for more shells

**Review by johnslavik — COMMENTED** [review]
Thanks!

File: docs/usage.md [hunk 2]
```diff
@@ -882,7 +882,15 @@ You can enable auto-completion in Bash with these commands:
 ```bash
 completions_dir="${BASH_COMPLETION_USER_DIR:-${XDG_DATA_HOME:-$HOME/.local/share}/bash-completion}/completions"
 mkdir -p "${completions_dir}"
-duty --completion > "${completions_dir}/duty"
+duty --completion=bash > "${completions_dir}/duty"
 ```
 
-Only Bash is supported for now.
+Or in Zsh with:
```
Comment (bswck) [inline docs/usage.md:888]: How about we use [content tabs](https://squidfunk.github.io/mkdocs-material/reference/content-tabs/#linked-content-tabs) instead?

File: src/duty/cli.py (hunk 1 above)
Comment (j-g00da) [reply src/duty/cli.py:291]: This is much better. 
Regarding `COMPLETION_DIR` as env var, I don't really see the benefit, can you elaborate on some use-cases?

Comment (pawamoy) [reply src/duty/cli.py:291]: Hmm `COMPLETION_DIR` here is just the folder containing our Python modules. It's therefore not configurable, and we can remove it from the message :thinking: Also it's fine to leave the declaration here, it won't be used anywhere else (I think) so no need to have it as a global variable.

File: src/duty/completions.zsh [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# Based on pyinvoke implementation of zsh completion
+_complete_duty() {
+    reply=( $(duty --complete -- ${words}) )
+}
+
+compctl -K _complete_duty + -f duty
```
[the remark this answers is not in the document]
Comment (j-g00da) [reply src/duty/completions.zsh:6]: Yes, tested on zsh 5.9

File: docs/usage.md [hunk 4]
```diff
@@ -882,7 +882,15 @@ You can enable auto-completion in Bash with these commands:
...
+
+```zsh
+completions_dir="$HOME/.duty"
```
Comment (pawamoy) [inline docs/usage.md:891]: I understand you may have taken inspiration from the Bash example above, but this doesn't seem right to me. The Bash example uses standard locations that Bash immediately understands. This location would require users to fiddle with the Zsh configuration to load the completion from `~/.duty` I believe. Furthermore, not using standard locations is generally frowned-upon by users (me included :smile:) as that leads to a cluttered HOME directory :slightly_smiling_face: 

In short, could you try to see if there's a standard location we could write the file in, so that Zsh's completion system natively finds it?

**Review by pawamoy — COMMENTED** [review]
Thanks a lot @j-g00da! It's really cool to get completions for Zsh too :smile: 

One thing I'd like to see for Zsh completions is actual description of the completion words: Zsh's completion system allows to attach a description to each term/word, that is then displayed when hitting TAB in the shell. That means we'd have to add a bit of code to our logic for generating word candidates here, so that descriptions are returned too. I see two options:

- the logic *always* returns tuples (word, description), and higher-up in the stack we filter out descriptions if the shell doesn't support those
- the logic accepts a `shell` argument that lets it know whether it should return descriptions as well as words

We would have to check how this integrates with the actual completion script (`completions.zsh`) and `compctl` command.

File: src/duty/cli.py (hunk 1 above)
Comment (bswck) [reply src/duty/cli.py]: It's just a loose idea from my intuition, without a strong argument to it.

This level of configurability can be a huge overkill, sure.

But in my perfect world, every software examining a path for presence of a concrete file would operate similarly to the `PATH` environment variable: the file is looked for subsequently in a bunch of directories.

In this particular case, a real use case would be to allow to fallback to user-defined directories to look up their custom-written completions.

`EXTRA_COMPLETION_DIRS=duty-completions/.` in a project with `duty-completions/` folder.

But yeah, I don't think I can see a person in 100 years that would do it, so feel free to not take this into account.

Also, maybe we should extend the error message even further to mention what it looked for:
```suggestion
        COMPLETION_DIR = Path(__file__).parent
        script_name = f"completions.{shell}"
        try:
            print((COMPLETION_DIR / script_name).read_text())
        except FileNotFoundError as exc:
            msg = f"Completion script not found for {shell!r}, looked for {script_name!r} in {COMPLETION_DIR}"
            raise NotImplementedError(msg) from exc
```

Comment (pawamoy) [reply src/duty/cli.py]: > In this particular case, a real use case would be to allow to fallback to user-defined directories to look up their custom-written completions, like EXTRA_DUTY_COMPLETION_DIRS=duty-completions/ in a project with duty-completions/ folder.

But... the `--completion` flag is used to *output* the completion script (that the user can then redirect to a file). If the user already has custom completion scripts somewhere, why would they re-output them through duty?

IMO we shouldn't even say "Completion script not found" as this implies a potential error. Instead we should say "Completions for {shell} shell are not available, feature requests and PRs welcome!".

File: docs/usage.md (hunk 4 above)
Comment (j-g00da) [reply docs/usage.md:891]: > In short, could you try to see if there's a standard location we could write the file in, so that Zsh's completion system natively finds it?

From what I understand, zsh doesn't have any "standard" location for completions other than `/usr/local/share/zsh/site-functions`, which is in fpath by default, but I don't think we want to install it system-wide. I based my approach on oh my zsh, which puts completions under `~/.oh-my-zsh/completions` and adds it to `fpath`. Still, I think cluttered HOME is a good point. Let me know what you think.

> One thing I'd like to see for Zsh completions is actual description of the completion words

Good point, I will work on this later today.

File: src/duty/cli.py (hunk 1 above)
Comment (bswck) [reply src/duty/cli.py]: Oops, I haven't seen https://github.com/pawamoy/duty/pull/34#discussion_r1942849094 before https://github.com/pawamoy/duty/pull/34#discussion_r1942870707.

> But... the `--completion` flag is used to output the completion script (that the user can then redirect to a file). If the user already has custom completion scripts somewhere, why would they re-output them through duty?

Yeah, that would be pointless.

File: docs/usage.md (hunk 4 above)
Comment (pawamoy) [reply docs/usage.md:891]: Thanks! We could then document that and let the user decide. I found https://github.com/zsh-users/zsh-completions/blob/master/zsh-completions-howto.org#telling-zsh-which-function-to-use-for-completing-a-command to be very readable, maybe we could link to it.

File: src/duty/cli.py (hunk 1 above)
Comment (bswck) [reply src/duty/cli.py]: > IMO we shouldn't even say "Completion script not found" as this implies a potential error. Instead we should say "Completions for {shell} shell are not available, feature requests and PRs welcome!".

Definitely, a good CTA is never bad!

Comment (j-g00da) [reply src/duty/cli.py]: 152040259ba68317798a6a03b270c712e51e885f

File: docs/usage.md (hunk 4 above)
Comment (j-g00da) [reply docs/usage.md]: I've modified the docs yestarday and changed the `completions.zsh` to use `compdef` (rich completions) instead of old `compctl`. Still no descriptions, but now it will be possible to add them. Will work on adding descriptions later today or tomorrow. You can take a look at the docs and let me know if it's enough.
Sources: [this answer on stackoverflow](https://stackoverflow.com/a/67161186), and ofc the [zsh-completions-howto.org](https://github.com/zsh-users/zsh-completions/blob/master/zsh-completions-howto.org).

**Comment by bswck** [comment]
I'm yearning for some automatic tests of those scripts. WDYT @pawamoy

File: docs/usage.md (hunk 4 above)
Comment (j-g00da) [reply docs/usage.md]: Also - completions.zsh is WIP, it's just a minimal example for now, but it works. 
Also 2 - We can make it compatible with old zsh versions, but I don't know if this is that important since user can always use bash completions in zsh if native approach doesn't work (explained in the docs).

Comment (pawamoy) [reply docs/usage.md]: Thank you so much! Don't bother about supporting old Zsh versions, latest is fine :slightly_smiling_face:

**Comment by pawamoy** [comment]
@bswck yes, testing is always good! I'm definitely not sure how to test auto-completion though! Any idea?

**Comment by bswck** [comment]
> Wouldn't it be enough to just test the output of our Python code that generates words?

Definitely no, I strongly feel we need white-box tests for completion working on the user end.

> I'm definitely not sure how to test auto-completion though! Any idea?

We can get some inspiration from [typer tests](https://github.com/fastapi/typer/tree/master/tests/test_completion) which seem to do exactly that.

**Comment by j-g00da** [comment]
I think of doing a bit of extra work on it, since the docs on how to enable completions grew in the Zsh section.

**Duty is simple, and installing completions for it should also be simple.** It’s fine to inform the user about options, but for the most part… it's too much and I really would like something like `duty --install-completions` that would just install it in some default place, for auto-detected shell so I don't have to think about it.

For Bash, it would be really easy - just run what’s in the docs. For Zsh I think the best option would be to put it in `/usr/local/share/zsh/site-functions`, because:
1. **This is what most tools use** - homebrew, uv, cargo to name a few.
2. **Doesn't require adding to `fpath`, as it is there by default**; which normally would be tricky, as `compinit` must be run _after_ adding a dir to `fpath`, so we would have to parse `.zshrc` and do some unholy stuff or ask user to manually do it, which misses the point.

In fact, I see only one disadvantage of such an approach - `/usr/local/share` is by default owned by root on linux. This is not a problem on macos, where `/usr/local/share` is owned by the current user. So, in short - this command would require `sudo` on Linux. I still think it's 100% worth it, let me know what you think @pawamoy, @bswck.

**Comment by j-g00da** [comment]
Also - I just checked Typer source code, since @bswck linked it, and this is exactly what they do:
* `--install-completion`: Install completion for the current shell.
* `--show-completion`: Show completion for the current shell, to copy it or customize the installation.

But when it comes to installation dir, it's `~/.zfunc`. So as I said, there really is no "default" directory for that in Zsh...
[source](https://github.com/fastapi/typer/blob/b8bac792c8c984df4c93777221b4cebd2dbe81b8/typer/_completion_shared.py#L124C1-L125C1):

```python
def install_zsh(*, prog_name: str, complete_var: str, shell: str) -> Path:
    # Setup Zsh and load ~/.zfunc
    zshrc_path = Path.home() / ".zshrc"
    zshrc_path.parent.mkdir(parents=True, exist_ok=True)
    zshrc_content = ""
    if zshrc_path.is_file():
        zshrc_content = zshrc_path.read_text()
    completion_line = "fpath+=~/.zfunc; autoload -Uz compinit; compinit"
    if completion_line not in zshrc_content:
        zshrc_content += f"\n{completion_line}\n"
    style_line = "zstyle ':completion:*' menu select"
    # TODO: consider setting the style only for the current program
    # style_line = f"zstyle ':completion:*:*:{prog_name}:*' menu select"
    # Install zstyle completion config only if the user doesn't have a customization
    if "zstyle" not in zshrc_content:
        zshrc_content += f"\n{style_line}\n"
    zshrc_content = f"{zshrc_content.strip()}\n"
    zshrc_path.write_text(zshrc_content)
    # Install completion under ~/.zfunc/
    path_obj = Path.home() / f".zfunc/_{prog_name}"
    path_obj.parent.mkdir(parents=True, exist_ok=True)
    script_content = get_completion_script(
        prog_name=prog_name, complete_var=complete_var, shell=shell
    )
    path_obj.write_text(script_content)
    return path_obj
```
On a side note they might have a potential bug there...

**Comment by pawamoy** [comment]
Completely agree with your [comment](https://github.com/pawamoy/duty/pull/34#issuecomment-2646209045) @j-g00da. I was exactly going to say that the only downside is that users will probably have to use sudo. IMO that's an acceptable tradeoff. If they don't want to use sudo, they can deal with their own configuration by writing the completion script where they prefer.

OK so it looks like Typer *modifies* `.zshrc` (in addition to cluttering home with a non-standard `.zfunc` directory it seems) :nauseated_face: Please let us not do that. Lets go with the approach you suggested:

- an `--install-completions` flag detects the shell (and accepts the shell as argument for user control)
- for Bash it uses the example we currently have
- for Zsh it tries to write in `/usr/local/share/zsh/site-functions`, and the docs say that one must probably run the command with sudo

**Comment by pawamoy** [comment]
@bswck @j-g00da do you know if Python has any standard or discussion about distributing shell completions inside packages (source dists or wheels)? Looks like venvs have the opportunity to store a, I don't know, `completions` folder somewhere. Then tool installers like pipx and uv could expose command to easily locate / load such completions. Each package could distribute completions for all (supported) shells, so that users actually never have to run commands to either install completions or redirect completions. CLI would be free of such flags.

Basher offers such a mechanism for example. Your project defines completion files for Bash, Zsh, whatever, and Basher puts them in a dedicated folder. Users can then easily point their shell at these folders.

**Comment by pawamoy** [comment]
I opened https://github.com/pypa/pipx/issues/1604 and https://github.com/astral-sh/uv/issues/11354, let see what people think :blush:

**Comment by bswck** [comment]
> I opened [pypa/pipx#1604](https://github.com/pypa/pipx/issues/1604) and [astral-sh/uv#11354](https://github.com/astral-sh/uv/issues/11354), let see what people think 😊

That would also fix https://github.com/pawamoy/git-changelog/issues/66#issuecomment-1889402715

**Comment by j-g00da** [comment]
> EDIT: PDM also suggests writing in ~/.zfunc, but I can't find any official docs about such a directory, so it looks like a convention?

Poetry suggests `.zfunc` too, but also lists omz directory as a good option.

Still - it's not backed by any standard AFAIK (it's often described as example of **custom** directory for completions) and because of that, the directory is not by default in `fpath` and requires adding it manually in `.zshrc`. Maybe I'm too deep into this, but I still have strong feelings that `site-functions` would be the best place for that. On the other hand, since it's popular, we can use `.zfunc` as an example of custom dir instead of a random `.duty`. Let me know what you think.

I still want to take a look into uv source code later, on my mac it installed completions in `site-functions`, I want to know if it does the same for linux.

**Comment by j-g00da** [comment]
> * the logic _always_ returns tuples (word, description), and higher-up in the stack we filter out descriptions if the shell doesn't support those

I went with this approach in 592f938b8cae108e122c6b8b9e2a601d38a0bf9d

image

**Comment by pawamoy** [comment]
Awesome :heart_eyes:

File: src/duty/_completion.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+    @staticmethod
+    def _zsh(candidates: list[CompletionCandidateType]) -> str:
+        def parse_candidate(item: CompletionCandidateType) -> str:
+            completion, help_text = item
+            # We only have space for one line of description,
+            # so I remove descriptions of sub-command parameters from help_text
+            # by removing everything after the first newline.
+            # I don't think it is the best approach and should be discussed.
+            return f"{completion}: {help_text or '-'}".split("\n", 1)[0]
```
Comment (pawamoy) [inline src/duty/_completion.py:35]: - functions docstrings: good, it's reasonable to keep only the first line, as docstrings' first line should be a complete sentence anyway.
- parameter docstrings: we don't parse them yet anyway (from looking at code below). When we do (using Griffe?), I don't think there's a more robust way than keeping the first line anyway. We don't want to enter natural language processing territory (splitting sentences is hard). Maybe we could recommend to use single-line descriptions for parameters, or at least to have a complete sentence as first line.

File: src/duty/collection.py [hunk 6]
```diff
@@ -157,14 +161,16 @@ def completion_candidates(self, args: tuple[str, ...]) -> list[str]:
...
+            (name, self.get(name).description or None) for name in names
+        )
 
         # If no duty found, return names.
         if name is None:
             return completion_names
 
         params = [
-            f"{param.name}="
+            (f"{param.name}=", None)
```
Comment (pawamoy) [inline src/duty/collection.py:173]: We could either parse docstrings here, with Griffe, or earlier, when loading duties (storing param descriptions in `Duty` instances). But we don't have to do this in this PR, can be done in a subsequent one :slightly_smiling_face: Just writing out the possibilities.

Comment (j-g00da) [reply src/duty/collection.py:173]: Yes, I think this can be made in a separate PR.

File: src/duty/completion.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
+"""Shell completion utilities."""
+
+import os
+import subprocess
+import sys
+from pathlib import Path
+from typing import Optional
+
+CompletionCandidateType = tuple[str, Optional[str]]
```
Comment (bswck) [inline src/duty/completion.py:9]:
```suggestion

if sys.version_info >= (3, 10):
    from typing import TypeAlias
else:
    from typing_extensions import TypeAlias
    
CompletionCandidateType: TypeAlias = "tuple[str, str | None]"
```

File: src/duty/completion.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+            # We only have space for one line of description,
+            # so we remove descriptions of sub-command parameters from help_text
+            # by removing everything after the first newline.
+            return f"{completion}: {help_text or '-'}".split("\n", 1)[0]
+
+        return "\n".join(parse_candidate(candidate) for candidate in candidates)
+
+    @staticmethod
+    def _bash(candidates: list[CompletionCandidateType]) -> str:
+        return "\n".join(completion for completion, _ in candidates)
```
Comment (bswck) [inline src/duty/completion.py:48]: So hacky! Patterns that involve unending static methods are usually meant to be in functional paradigm :)
Btw https://typing.readthedocs.io/en/latest/reference/best_practices.html#arguments-and-return-types
But looking at this PR in a wider perspective, my suggestion will be out of scope of this class only.

File: src/duty/completion.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+
+    @classmethod
+    def install(cls, shell: str) -> None:
+        """Installs shell completions for selected shell.
+
+        Raises:
+            NotImplementedError: When installer is not implemented for selected shell.
+        """
+        try:
+            return getattr(cls, f"_{shell}")()
```
Comment (bswck) [inline src/duty/completion.py:62]: This is avoidable by interface segregation instead of fine-grained handlers per topic.

File: src/duty/completion.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+        except PermissionError:
+            # retry as sudo
+            if os.geteuid() == 0:
+                raise
+            subprocess.run(  # noqa: S603
+                ["sudo", sys.executable, sys.argv[0], "--install-completion=zsh"],  # noqa: S607
+                check=True,
+            )
+        except FileExistsError:
+            print("Zsh completions already installed.")
```
Comment (bswck) [inline src/duty/completion.py:96]: Better to write to stderr if ever reporting an error :)

**Review by johnslavik — CHANGES_REQUESTED** [review]
Let's group behaviors by shells instead of topics.

Instead of fine-grained repetitive interfaces

`topic1`
- `bash`
- `zsh`
- _other shells..._

`topic2`
- `bash`
- `zsh`
- _other shells..._

we want modular strategies, like

`bash`
- `topic1`
- `topic2`

`zsh`
- `topic1`
- `topic2`

_other shells follow the same pattern..._

This approach is known under many names; the _I from SOLID_, the _strategy design pattern_, the _loose coupling_, the _high cohesion_.

Notice that the current solution doesn't really leverage the use of classes for grouping methods by common state; an interface for a "shell" can be easily described with an abstract class/protocol, distinct implementations gathered within a class-level registry modified via `__init_subclass__` of the abstract class/protocol. That would cause future implementations of other shells to be cohesive, discoverable, comprehensive and complete, with a potential of sharing useful state across "topic" methods; something along the lines of:

```py
class Shell(metaclass=abc.ABCMeta):
    name: ClassVar[str]
    implementations: Final[ClassVar[dict[str, Shell]]] = {}

    @abc.abstractmethod
    def parse_completions(self, candidates: Sequence[CompletionCandidateType]) -> str:
        ...

    @abc.abstractmethod
    def install_completions(self, candidates: Sequence[CompletionCandidateType]) -> str:
        ...

    def __init_subclass__(cls) -> None:
        cls.implementations[cls.name] = cls


class Bash(Shell):
    name = "bash"

    def parse_completions(self, candidates: Sequence[CompletionCandidateType]) -> str:
        # implementation for bash

    def install_completions(self) -> str:
        # implementation for bash


class Zsh(Shell):
    name = "zsh"

    def parse_completions(self, candidates: Sequence[CompletionCandidateType]) -> str:
        # implementation for zsh

    def install_completions(self) -> str:
        # implementation for zsh


# other shells follow the same pattern
```

Implementing a new shell/changing an existing shell implementation will only require changes in one area in the code, instead of a number of them, all over the place, separated by other, unrelated fragments.
Notice how this approach makes the new implementations also portable and easier to plug into the pre-existing runtime—simply implement a strategy for your shell and ship it. The library can find it in the `Shell.implementations` registry and you can override the existing ones as well, and then maybe suggest a patch to upstream. 

I'm up for pair programming on this one, hit me up on Discord if you're interested :)

**Comment by j-g00da** [comment]
Thanks for feedback @bswck. Generally there are only tests left to do now @pawamoy. I'll do some research and experiment with it today. There are still some things worth pointing out though:
1. The `$SHELL` check I'm doing when no shell is provided is very error prone. I think we can just use a parameter in docs example to encourage users to provide it, the help text already points out that `$SHELL` is used when no param provided (or maybe we should just require it?)
2. I already talked with @bswck about it - I used prints to be consistent with the rest of the codebase, but maybe we should create an issue to use a proper logger?
3. Do we want to test for backward compatibility of shell completion scripts? Eg. I had to change `completions.bash` and make sure the old script will still work when someone upgrades from 1.5.0 and has bash completions installed already (without manually deleting the old script from the system and reinstalling completions.) Regular unit tests wouldn't catch that.

**Comment by pawamoy** [comment]
Thank you so much for all your hard work on this @j-g00da :pray: 

1. We can keep the frail logic, but indeed recommend using the `--shell` option, and always use it in the docs. I'm curious though, what makes you say it's error-prone?
2. Let's stay with `print` for now. I've not made up my mind yet on a way to use and configure logging for both logging and terminal display. Happy to hear about your own experience on that matter though!
3. Good question. We definitely do not want completions to break with ugly errors when users update duty. I think that as long as we respect the completion interface of the supported shells, and that we don't make breaking changes in the CLI (which is called by the completion scripts), we should be fine. For now the only backward compatibility precaution you added is us defaulting to `bash` for `--complete`, right? Do you see other things that would need such backward compatibility precautions right now or in the near future? In any case, it's easy enough for users to just re-install completions when upgrading duty (at least if/when they get an error). We just have to document this prominently. Also, if/when pipx/uv start both supporting man pages and completion scripts, this point becomes moot, as completions will always be up-to-date.

**Comment by j-g00da** [comment]
> I'm curious though, what makes you say it's error-prone?

`$SHELL` is user shell, not current shell - if you for example launch bash from zsh, it will still say `/bin/zsh`.

**Comment by j-g00da** [comment]
> Do you see other things that would need such backward compatibility precautions right now or in the near future?

There are some, but since we are using symlinks (in `--install-completion`) - as long as paths to completion scripts are the same, they should update automatically. So the main issue can happen if we move/rename `completions.*` files.

On the other hand should we really care if completion scripts were installed manually? Sure, we can make test specifically for backward compatibility with 1.5.0, as currently this is the only way to install completions, but with the next update we can promote usage of `--install-completion` and add warning in docs, that if you choose to install completions manually, you are on your own. I think that's fair and then we can just make a test that checks if `completions.*` exist for each shell in proper location...

WDYT? @pawamoy @bswck

**Comment by pawamoy** [comment]
> So the main issue can happen if we move/rename completions.* files.

Hmm.

- Symlinks: auto-updates, work as long as path stay the same (both the completion file path within our project, and the installation path on the user machine, which depends on the Python version! reinstall same duty version on different Python version and the symlink will be broken)
- Hard links: depending on how pip updates work, inode could change, separating hard link from source file, meaning no auto-update. Uninstall and reinstalling would isolate the hard link for sure.

:thinking: 

In the end it looks to me like we should always recommend users to update the completion script upon updating duty. Also, if the file already exists, don't error out, force rewrite it. I'm waiting for the PR to get out of draft for a full review :slightly_smiling_face:

**Comment by j-g00da** [comment]
Just pushed some work in progress (only bash for now) tests to show what I'm trying to do here. Since some of these tests modify the system, I decided to mark these tests with custom pytest mark and run them in isolated containers. Just a proof of concept for now, not optimized, I'm still not entirely sure if it's the way to go.

File: tests/test_completion/test_bash.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+
+
+@parametrize_completions
+@needs_shell(Bash)
+@needs_platform("linux", "darwin")
+@needs_isolated_container
+@pytest.mark.isolate
+def test_completion_function(duties_file: str, partial: str, expected: str) -> None:
+    """Test bash `_complete_duty` function."""
+    # TODO: Temporary hack, as for now completions don't respect the `-d` flag - to be fixed in another PR.
```
Comment (j-g00da) [inline tests/test_completion/test_bash.py:118]: Also a side note - completions always use duties.py in main directory to generate candidates because `-d` param is not passed to completion script. I can try to fix it when I'm done with tests, but maybe let's make it another PR.

File: .github/workflows/ci.yml [hunk 12]
```diff
@@ -55,6 +55,92 @@ jobs:
...
+    - name: Install dependencies
+      env:
+        UV_RESOLUTION: ${{ matrix.resolution }}
+      run: make setup
+
+    - name: Run the test suite
+      env:
+        _DUTY_ISOLATED_TEST_CONTAINER: true
+      run: duty test ${{ matrix.pytest-nodeid }} parallel=False
+
```
Comment (pawamoy) [inline .github/workflows/ci.yml:143]: Can't we just add a single test-completions job, using pytest's `-k` option, exposed as `match` parameter on the `test` duty? IIUC the current solution would create lots of jobs, one per collected test. Also, lots of boilerplate :confused:

Comment (pawamoy) [reply .github/workflows/ci.yml:143]: We could use pytest-xdist's `--dist=loadfile` option, to make sure all tests within a given module are executed by the same worker. Each test would clean up after itself. I believe this would prevent any race condition.

Comment (j-g00da) [reply .github/workflows/ci.yml:143]: > IIUC the current solution would create lots of jobs.

Yes, that's why I'm not really sure about it. On the other hand this is the only way I see to make these tests deterministic. If we run multiple tests on the same machine, we can create some cleanup mechanism but it won't be failproof and will depend on implementation - and if a change in implementation needs a change in corresponding test, then it's not a good test.

Also there is problem of parallelism as you said. Even if we run them in one job (per python version and os), we can't parallelize it, so we still need a second matrix for them.

Boilerplate can be reduced for sure, I just wanted to share what I'm trying to do, because this needs to be discussed.

Comment (pawamoy) [reply .github/workflows/ci.yml:143]: > and if a change in implementation needs a change in corresponding test, then it's not a good test.

That's a tradeoff I'm willing to accept.

> easiest solution would be to just call {Shell}().install_path.unlink()

Sounds very reasonable to me :slightly_smiling_face:

Comment (pawamoy) [reply .github/workflows/ci.yml:143]: I mean, yeah, integration tests are nice to have (detecting early when completion install commands do not work as expected anymore), but at the same I don't expect shells to change frequently (and that's an euphemism). So I'm on the side of "lets not waste energy writing and maintaining integration tests that aren't relevant to the project specifically".

Comment (j-g00da) [reply .github/workflows/ci.yml:143]: Sure, I will get back to this, when I have some free time. I think we can also force pytest to run the installation test first (before edge cases like overriding previous installation.)

Comment (pawamoy) [reply .github/workflows/ci.yml:143]: I would not recommend forcing the order of tests as this can hide dependencies between them. Random order forces us to make sure each test is atomic and complete and correctly cleans up after itself.

And of course, no rush, thank you so much for all your work already :smile:

Comment (j-g00da) [reply .github/workflows/ci.yml:143]: Good point!

File: docs/usage.md (hunk 4 above)
Comment (pawamoy) [reply docs/usage.md]: Resolving since the code is outdated.

File: src/duty/_completion.py (hunk 5 above)
Comment (pawamoy) [reply src/duty/_completion.py]: Resolving (outdated code).

**Comment by pawamoy** [comment]
Just created this discussion: https://discuss.python.org/t/standard-path-in-wheels-for-completion-scripts-and-manpages/106396.

**Comment by pawamoy** [comment]
@j-g00da I've taken some time to finish the work started in this PR. 

I kinda messed up my Git stuff and ended up pushing from a local branch to the remote main. @j-g00da you're a co-author of the commit, see https://github.com/pawamoy/duty/commit/50e9cfc34ed98b6143e9947af34b8690824e0cdb. Let me know if there's something wrong or something you'd like me to change, I didn't release anything yet, we can force-push :slightly_smiling_face:

**Comment by j-g00da** [comment]
looks good :)
